p503/arith_decl.go

```diff
@@ -0,0 +1,87 @@
+// +build amd64,!noasm
+
+package p503
+
+import (
+	. "github.com/cloudflare/p751sidh/internal/isogeny"
+	"golang.org/x/sys/cpu"
+)
+
+// If choice = 0, leave x,y unchanged. If choice = 1, set x,y = y,x.
+// If choice is neither 0 nor 1 then behaviour is undefined.
+// This function executes in constant time.
+//go:noescape
+func fp503ConditionalSwap(x, y *FpElement, choice uint8)
+
+// Compute z = x + y (mod p).
+//go:noescape
+func fp503AddReduced(z, x, y *FpElement)
+
+// Compute z = x - y (mod p).
+//go:noescape
+func fp503SubReduced(z, x, y *FpElement)
+
+// Compute z = x + y, without reducing mod p.
+//go:noescape
+func fp503AddLazy(z, x, y *FpElement)
+
+// Compute z = x + y, without reducing mod p.
+//go:noescape
+func fp503X2AddLazy(z, x, y *FpElementX2)
+
+// Compute z = x - y, without reducing mod p.
+//go:noescape
+func fp503X2SubLazy(z, x, y *FpElementX2)
+
+// Reduce a field element in [0, 2*p) to one in [0,p).
+//go:noescape
+func fp503StrongReduce(x *FpElement)
+
+// Function pointer to function computing z = x * y.
+// Concrete implementation depends on capabilities of the CPU which
+// are resolved at runtime. CPUs with ADCX, ADOX and MULX support
+// run most optimized implementation
+var fp503Mul func(z *FpElementX2, x, y *FpElement)
+
+// Mul implementattion for legacy CPUs
+//go:noescape
+func mul(z *FpElementX2, x, y *FpElement)
+
+// Mul implementation for CPUs supporting two carry chains
+//go:noescape
+func mulWithMULX(z *FpElementX2, x, y *FpElement)
+
+// Mul implementation for CPUs supporting two carry chains and MULX multiplayer
+//go:noescape
+func mulWithMULXADX(z *FpElementX2, x, y *FpElement)
+
+// Computes the Montgomery reduction z = x R^{-1} (mod 2*p). On return value
+// of x may be changed. z=x not allowed.
+var fp503MontgomeryReduce func(z *FpElement, x *FpElementX2)
+
+func redc(z *FpElement, x *FpElementX2)
+
+// Mul implementation for CPUs supporting two carry chains and MULX multiplayer
+//go:noescape
+func redcWithMULX(z *FpElement, x *FpElementX2)
+
+// Mul implementation for CPUs supporting two carry chains
+//go:noescape
+func redcWithMULXADX(z *FpElement, x *FpElementX2)
+
+// On initialization, set the fp503Mul function pointer to the
+// fastest implementation depending on CPU capabilities.
+func init() {
+	if cpu.X86.HasBMI2 {
+		if cpu.X86.HasADX {
+			fp503Mul = mulWithMULXADX
```

Comment:
Actually this is more expensive than having wrapper function with an if. It may be possible to do some kind of late binding at runtime, but it's going to be hacky

Reply:
This will be done in next PR together with changes to P751